File: 42.接雨水.py
#
# @lc app=leetcode.cn id=42 lang=python3
#
# [42] 接雨水
#

# @lc code=start
from typing import List
from functools import lru_cache
class Solution:
    @lru_cache
    def lmax(self,x:int)->int:
        if x==0:return self.height[x]
        return max(self.lmax(x-1),self.height[x])
    @lru_cache
    def rmax(self,x:int)->int:
        if x==len(self.height)-1:return self.height[x]
        return max(self.height[x],self.rmax(x+1))
    def trap(self, height: List[int]) -> int:
        self.height=height
        ret=0
        for i in range(1,len(self.height)-1):
            ret+=min(self.lmax(i),self.rmax(i))-self.height[i]
        return ret
# @lc code=end

File: 42.接雨水.cpp
/*
 * @lc app=leetcode.cn id=42 lang=cpp
 *
 * [42] 接雨水
 */

// @lc code=start
#include<vector>
using namespace std;
class Solution {
public:
    int trap(vector<int>& height) {
        int n = height.size(), ret = 0;
        vector<int> left_max(n), right_max(n);
        left_max.front() = height.front();
        right_max.back() = height.back();
        for(int i=1; i<n; ++i){
            left_max[i] = max(left_max[i-1], height[i]);
            right_max[n-i-1] = max(right_max[n-i], height[n-i-1]);
        }
        for(int i=0; i<n; ++i){
            ret += min(left_max[i], right_max[i]) - height[i];
        }
        return ret;
    }
};
// @lc code=end

